### ISSUE TITLE
Exporting User Accounts

### ISSUE THREAD

**Issue by SzymonSel** [OrchardCMS/OrchardCore#13073]
I ended up here #7891 looking for a way of export and importing users from one istance to another. Is this still planned as a deployment task? What is the state of it's developement? Any way I could contribute?

**Comment by hishamco** [comment]
We need a secrets module to deploy any sort of secrets in more confidential way, then deployments & recipes will take care about the rest

@deanmarcussen any plan to complete the Secrets module that you started with? Or shall some one continues the work that you already did?

**Comment by hishamco** [comment]
Yes we could, but it would be nice to export/import users with their credentials

**Comment by sebastienros** [comment]
@deanmarcussen twice in a day ... Do you remember where you code is at if there is a way that someone takes over?

**Comment by hishamco** [comment]
FYI @sebastienros  #7891

**Comment by SzymonSel** [comment]
> Yes we could, but it would be nice to export/import users with their credentials

To achieve this now, what would one have to do? Where are the keys stored? How can they be copied to make the credentials work?

**Comment by hishamco** [comment]
That's what Secrets Module is all about

**Comment by sebastienros** [comment]
Do we actually need the secrets module to import/export users? Is there any state in the web app that makes changes how these hashes are stored in the database? I don't think so, so sending them should be fine. There could be an option to include the hashes, this way one could already export the list of user accounts (requires a custom permission though). And importing a user is "security sensitive", i.e. you could reset someone's password with this permission. So not sure if it's a good idea without encryption. Or at least with the same level of permissions as being able to set the encryption keys.

**Comment by hishamco** [comment]
> Do we actually need the secrets module to import/export users? 

Yes & No it depends, it's only need when we want to export content with sensitive data

> Is there any state in the web app that makes changes how these hashes are stored in the database? I don't think so, so sending them should be fine.

I think it's fine too send hashes too, but we need to think about more generic way for any sensitive data such as SMTP credentials, database connections .. etc

We might need to revise the secrets modules and make it done if it's possible

**Comment by sebastienros** [comment]
> but we need to think about more generic way for any sensitive data such as SMTP credentials

Read the title of this issue

**Comment by hishamco** [comment]
I already did, but look to the issue description :)

> I ended up here https://github.com/OrchardCMS/OrchardCore/pull/7891 looking for a way of export and importing users from one istance to another. Is this still planned as a deployment task? What is the state of it's developement? Any way I could contribute?

So, for this particular use case it's fine to export the users data

**Comment by sebastienros** [comment]
We checked the password hashing code and there is no secret there, so it's safe to create an export step for the user accounts.

**Comment by SzymonSel** [comment]
Great! I’m on it!
-- 

[image: Onefourseven Sp. z o.o.]

Szymon SELIGA
CEO | Hackeman
+48 505 232 294

Onefourseven Labs
01-793 Warszawa, ul. Rydygiera 8 bud. 10
www.onefourseven.pl

**Comment by yassinsayed** [comment]
@SzymonSel  did you find a way to import users

**Comment by SzymonSel** [comment]
I have, but I haven't yet implemented it.

**Comment by SzymonSel** [comment]
I have, but I haven't yet implemented it.

**Comment by SzymonSel** [comment]
I've succesfully implemented a simple Users' Account  Deployment step together with the importing step. There is yet no advanced validation or any permissions checking, but this could be implemented at a later time.

Should I place my implemetation in the OrchardCore.Users project for review?

**Comment by hishamco** [comment]
Sure

**Comment by SzymonSel** [comment]
I'm exporting the `PasswordHash`

**Comment by hishamco** [comment]
As expected, even though exporting such data is risky

/cc @sebastienros @Piedone

**Comment by jtkech** [comment]
I'm working on the secrets module which will use a combination of symetric and assymetric encryption.

**Comment by hishamco** [comment]
I see you are active to that PR which I planned to make some progress on it, so please assign the PR or the issue to you to avoid anyone else from duplicate the work

**Comment by sebastienros** [comment]
@SzymonSel I think it's safe to add the feature. Thanks a lot

**Comment by SzymonSel** [comment]
I've also created a simple `CustomUserSettings` Deployment implementation. 

By the way, can some one grant me push permissions?

**Comment by SzymonSel** [comment]
Sure thing! I'm being lazy and couldn't be bothered forking the repo :P

**Comment by Piedone** [comment]
Ah, I didn't know you implemented this, great!
